### PR TITLE
Add new test font

### PR DESCRIPTION
**Pull request by LongCatIsLooong**
_I'm planning to keep Ahem the default until the framework tests are migrated (https://github.com/flutter/flutter/pull/121306)._

Add a new test font that provides accurate font & glyph metrics ~and some Chinese character codepoint coverage~ (not sure if that's still needed).

Ahem's 1000 upem (not a power of 2) causes FreeType to truncate numbers when computing metrics, since it will be used as a divisor and calculations are done using 26.6 fixed-point numbers. For example when `point size = 1`, FreeType would report an x advance of `0.99998` instead of `1`. Currently we round most of the horizontal metrics so there are tons of `expect(text.width, 100)` in framework tests. Those tests are going to fail once the rounding is removed.  
The new font covers most [codepoint ranges Ahem covers](https://www.w3.org/Style/CSS/Test/Fonts/Ahem/), with glyphs that look almost identical to glyphs from Ahem, except with a different baseline location (`ASCENT : DESCENT = 3 : 1`, since Ahem's ratio `ASCENT : DESCENT = 4 : 1` can't be preserved when `upem` is a power of 2).

Some font metrics: `ascent = 768`, `descent = 256`, `units per em = 1024`

### Glyph Mappings by Scripts

|     \ Script <br />Glyph | DFLT | grek | hani | latn |
 | :---  | :----:  | :----:  | :----:  | :----: |
| Square | **codepoint(s):** 0x21-0x26, 0x28-0x40, 0x5b-0x60, 0x7b-0x7e, 0xa1-0xa9, 0xab-0xb9, 0xbb-0xbf, 0xd7, 0xf7, 0x2c6-0x2c7, 0x2c9, 0x2d8-0x2dd, 0x2013-0x2014, 0x2018-0x201a, 0x201c-0x201e, 0x2020-0x2022, 0x2026, 0x2030, 0x2039-0x203a, 0x2044, 0x2122, 0x2202, 0x2206, 0x220f, 0x2211-0x2212, 0x2219-0x221a, 0x221e, 0x222b, 0x2248, 0x2260, 0x2264-0x2265, 0x22f2, 0x25ca, 0xf000-0xf002<br />**character(s):** `!` `"` `#` `$` `%` `&` `(` `)` `*` `+` `,` `-` `.` `/` `0` `1` `2` `3` `4` `5` `6` `7` `8` `9` `:` `;` `<` `=` `>` `?` `@` `[` `\` `]` `^` `_` `` ` `` `{` `\|` `}` `~` `¡` `¢` `£` `¤` `¥` `¦` `§` `¨` `©` `«` `¬` `<SOFT HYPHEN>` `®` `¯` `°` `±` `²` `³` `´` `µ` `¶` `·` `¸` `¹` `»` `¼` `½` `¾` `¿` `×` `÷` `ˆ` `ˇ` `ˉ` `˘` `˙` `˚` `˛` `˜` `˝` `–` `—` `‘` `’` `‚` `“` `”` `„` `†` `‡` `•` `…` `‰` `‹` `›` `⁄` `™` `∂` `∆` `∏` `∑` `−` `∙` `√` `∞` `∫` `≈` `≠` `≤` `≥` `⋲` `◊` `<0xf000>` `<0xf001>` `<0xf002>` | **codepoint(s):** 0x394, 0x3a5, 0x3a7, 0x3a9, 0x3bc, 0x3c0, 0x2126<br />**character(s):** `Δ` `Υ` `Χ` `Ω` `μ` `π` `Ω` | **codepoint(s):** 0x3007, 0x4e00, 0x4e03, 0x4e09, 0x4e2d, 0x4e5d, 0x4e8c, 0x4e94, 0x516b, 0x516d, 0x5341, 0x5426, 0x56d7, 0x56db, 0x571f, 0x6587, 0x6587, 0x662f, 0x6728, 0x672c, 0x6b63, 0x6c34, 0x6d4b, 0x706b, 0x786e, 0x8bd5, 0x91d1<br />**character(s):** `〇` `一` `七` `三` `中` `九` `二` `五` `八` `六` `十` `否` `囗` `四` `土` `文` `文` `是` `木` `本` `正` `水` `测` `火` `确` `试` `金` | **codepoint(s):** 0x41-0x5a, 0x61-0x7a, 0xaa, 0xba, 0xc0-0xc8, 0xca-0xd6, 0xd8-0xf6, 0xf8-0xff, 0x131, 0x152-0x153, 0x178, 0x192<br />**character(s):** `A` `B` `C` `D` `E` `F` `G` `H` `I` `J` `K` `L` `M` `N` `O` `P` `Q` `R` `S` `T` `U` `V` `W` `X` `Y` `Z` `a` `b` `c` `d` `e` `f` `g` `h` `i` `j` `k` `l` `m` `n` `o` `p` `q` `r` `s` `t` `u` `v` `w` `x` `y` `z` `ª` `º` `À` `Á` `Â` `Ã` `Ä` `Å` `Æ` `Ç` `È` `Ê` `Ë` `Ì` `Í` `Î` `Ï` `Ð` `Ñ` `Ò` `Ó` `Ô` `Õ` `Ö` `Ø` `Ù` `Ú` `Û` `Ü` `Ý` `Þ` `ß` `à` `á` `â` `ã` `ä` `å` `æ` `ç` `è` `é` `ê` `ë` `ì` `í` `î` `ï` `ð` `ñ` `ò` `ó` `ô` `õ` `ö` `ø` `ù` `ú` `û` `ü` `ý` `þ` `ÿ` `ı` `Œ` `œ` `Ÿ` `ƒ` |
| Ascent Flushed |  |  |  | **codepoint(s):** 0x70<br />**character(s):** `p` |
| Descent Flushed |  |  |  | **codepoint(s):** 0xc9<br />**character(s):** `É` |
| Full Advance | **codepoint(s):** 0x20<br />**character(s):** `<SPACE>` |  |  |  |
| 1/2 Advance | **codepoint(s):** 0x2002<br />**character(s):** `<EN SPACE>` |  |  |  |
| 1/3 Advance | **codepoint(s):** 0x2004<br />**character(s):** `<THREE-PER-EM SPACE>` |  |  |  |
| 1/4 Advance | **codepoint(s):** 0x2005<br />**character(s):** `<FOUR-PER-EM SPACE>` |  |  |  |
| 1/6 Advance | **codepoint(s):** 0x2006<br />**character(s):** `<SIX-PER-EM SPACE>` |  |  |  |
| 1/5 Advance | **codepoint(s):** 0x2009<br />**character(s):** `<THIN SPACE>` |  |  |  |
| 1/10 Advance | **codepoint(s):** 0x200a<br />**character(s):** `<HAIR SPACE>` |  |  |  |
| Zero Advance | **codepoint(s):** 0xfeff<br />**character(s):** `<ZERO WIDTH NO-BREAK SPACE>` |  |  |  |

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on writing and running engine tests.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I signed the [CLA].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by chinmaygarde** [comment]
cc @Hixie

**Comment by LongCatIsLooong** [comment]
(Still trying to figure out how web tests load fonts but other than that should be ready for review).

**Comment by tgucio** [comment]
@LongCatIsLooong I'm thinking maybe adding some code points in the U+10000+ range could be useful to test behaviour with surrogate pairs? Say, Musical Symbols Range: U+1D100–U+1D1FF or Mathematical Alphanumeric Symbols Range: U+1D400–U+1D7FF.

Other interesting items for tests (thinking of some of the existing issues around complex scripts in text fields):
- glyphs that will result in clusters in text shaping (e.g. Devanagari)
- GSUB, GPOS features that affect caret position

**Comment by LongCatIsLooong** [comment]
This font is meant for framework tests. In the framework we can tell whether a codepoint is in the supplementary planes by looking at the range and don't have to rely on the shaper so I think it's rarely needed to have glyph mappings for that. If you have specific needs feel free to update the script and regenerate the font (but Skia is likely a better place to write tests to verify the shapers are working as expected, the framework currently doesn't direct configure shapers AFAIK), or use `FontLoader` to load a suitable font.

**Comment by tgucio** [comment]
@LongCatIsLooong yes I meant framework tests for issues like 
https://github.com/flutter/flutter/issues/118403 (clusters from OpenType shaping) or maybe also https://github.com/flutter/flutter/issues/98516 (clusters from combining marks) where the shaper most likely works as expected but the framework doesn't handle caret location well. Still, I guess you're right this might be a topic for another PR.

For this one perhaps we could just add some > U+10000 characters to CMAP as that should be cheap.

**Review by mdebbar — APPROVED** [review]
Web changes look good to me with some minor suggestions!

**Comment by jason-simmons** [comment]
Overall approach LGTM

The tests are segfaulting on CI.  In particular, the runs of flutter_tester with ASAN are reporting a null pointer dereference in `flutter::GetTestFontData`.

File: web_sdk/BUILD.gn [hunk 1]
```diff
@@ -144,6 +145,13 @@ sdk_rewriter("web_unicode_library") {
   output_dir = "$root_out_dir/flutter_web_sdk/lib/_web_unicode/"
 }
 
+sdk_rewriter("web_test_fonts_library") {
+  library_name = "web_test_fonts"
+  api_file = "//flutter/third_party/web_test_fonts/lib/web_test_fonts.dart"
+  input_dir = "//flutter/third_party/web_test_fonts/lib/web_test_fonts/"
```
Comment (mdebbar) [inline web_sdk/BUILD.gn:151]: This package needs to exist. For example, look at:

- https://github.com/flutter/engine/tree/main/third_party/web_unicode
- https://github.com/flutter/engine/tree/main/third_party/web_locale_keymap

Perhaps you forgot to include them in the commit?

Comment (LongCatIsLooong) [reply web_sdk/BUILD.gn:142]: Sorry forgot to track those files. The tests are passing now.

**Comment by LongCatIsLooong** [comment]
@tgucio I think the issues you mentioned are likely skparagraph issues and it's better to add tests to skia, as in those cases it seems the framework isn't doing anything particularly wrong, and it's going to fail faster so should there be any regressions skia maintainers will notice that without having to roll the changes into flutter. Also for skparagraph it's easy to include/load font assets in their tests so you could use existing fonts.

**Comment by LongCatIsLooong** [comment]
I didn't realize we were running `flutter_tester` without flutter tool. I guess It's safer to embed the font data so I'll keep the existing fonts as they are and embed the new font too. It looks like asan is no longer complaining.

**Comment by LongCatIsLooong** [comment]
Looks like the tests are passing. Could you take another look at the new approach?

**Review by jason-simmons — APPROVED** [review]

**Review by mdebbar — APPROVED** [review]
LGTM!

**Pull request merged**